source/librii/rarc/RARC.cpp

```diff
@@ -0,0 +1,1281 @@
+  return true;
+}
+
+bool ReplaceNode(ResourceArchive& rarc, ResourceArchive::Node to_replace,
+                 const std::filesystem::path& src) {
+  if (!std::filesystem::exists(src))
+    return false;
+
+  auto node_it = std::find(rarc.nodes.begin(), rarc.nodes.end(), to_replace);
+  if (node_it == rarc.nodes.end())
+    return false;
+
+  if (to_replace.is_folder()) {
+    if (!std::filesystem::is_directory(src))
+      return false;
+
+    auto parent = std::find_if(rarc.nodes.begin(), rarc.nodes.end(),
+                               [&](ResourceArchive::Node node) {
+                                 return node.id == to_replace.folder.parent;
+                               });
+
+    auto begin = node_it;
+    auto end = rarc.nodes.begin() + to_replace.folder.sibling_next;
+    auto old_size = std::distance(begin, end);
+
+    auto deleted_at =
+        std::distance(rarc.nodes.begin(), rarc.nodes.erase(begin, end));
+
+    // Generate an archive so we can steal the DFS structure.
+    auto tmp_rarc = librii::RARC::CreateResourceArchive(src);
+    if (!tmp_rarc)
+      return false;
+
+    tmp_rarc->nodes[0].name = to_replace.name;
+    tmp_rarc->nodes[0].folder.parent =
+        to_replace.folder.parent; // This repairs the insertion node which will
+                                  // recursively regenerate later
+
+    for (auto& new_node : tmp_rarc->nodes) {
+      if (new_node.is_folder())
+        new_node.folder.sibling_next += deleted_at;
+    }
+    rarc.nodes.insert(rarc.nodes.begin() + deleted_at, tmp_rarc->nodes.begin(),
+                      tmp_rarc->nodes.end());
+
+    auto sibling_adjust = tmp_rarc->nodes.size() - old_size;
+    if (sibling_adjust != 0) {
+      for (auto node = rarc.nodes.begin(); node != rarc.nodes.end();) {
+        auto index = std::distance(rarc.nodes.begin(), node);
+        // Skip files and the special dirs (they get recalculated later)
+        if (!node->is_folder()) {
+          node++;
+          continue;
+        }
+        if (index == deleted_at) { // Skip inserted dir (already adjusted)
+          node = rarc.nodes.begin() + node->folder.sibling_next;
+          continue;
+        }
+        if (node->folder.sibling_next <
+            deleted_at) { // Skip dirs that come before insertion by span
+                          // alone
+          node = rarc.nodes.begin() + node->folder.sibling_next;
+          continue;
+        }
+        // This is a directory adjacent but not beyond the insertion, ignore.
+        if (node->folder.parent >= parent->id && index < deleted_at) {
+          node = rarc.nodes.begin() + node->folder.sibling_next;
+          continue;
+        }
+        node->folder.sibling_next += sibling_adjust;
+        node++;
+      }
+    }
+
+    return true;
+  }
+
+  if (!std::filesystem::is_regular_file(src))
+    return false;
+
+  auto in = std::ifstream(src.string(), std::ios::binary | std::ios::in);
+  auto fsize = std::filesystem::file_size(src);
```

Comment:
I think all of these filesystem calls can throw exceptions, so it might be good to pass a `std::errc&` reference to use the no-throw version.